src/access.rs

```diff
@@ -0,0 +1,325 @@
+	D: BatchDatabase,
+{
+	blockchain: EsploraBlockchain,
+	wallet: Mutex<bdk::Wallet<D>>,
+	queued_transactions: Mutex<Vec<Txid>>,
+	watched_transactions: Mutex<Vec<Txid>>,
+	queued_outputs: Mutex<Vec<WatchedOutput>>,
+	watched_outputs: Mutex<Vec<WatchedOutput>>,
+	last_sync_height: Mutex<Option<u32>>,
+	logger: Arc<FilesystemLogger>,
+}
+
+impl<D> LdkLiteChainAccess<D>
+where
+	D: BatchDatabase,
+{
+	pub(crate) fn new(
+		blockchain: EsploraBlockchain, wallet: bdk::Wallet<D>, logger: Arc<FilesystemLogger>,
+	) -> Self {
+		let wallet = Mutex::new(wallet);
+		let watched_transactions = Mutex::new(Vec::new());
+		let queued_transactions = Mutex::new(Vec::new());
+		let watched_outputs = Mutex::new(Vec::new());
+		let queued_outputs = Mutex::new(Vec::new());
+		let last_sync_height = Mutex::new(None);
+		Self {
+			blockchain,
+			wallet,
+			queued_transactions,
+			watched_transactions,
+			queued_outputs,
+			watched_outputs,
+			last_sync_height,
+			logger,
+		}
+	}
+
+	pub(crate) fn sync_wallet(&self) -> Result<(), Error> {
+		let sync_options = SyncOptions { progress: None };
+
+		self.wallet
+			.lock()
+			.unwrap()
+			.sync(&self.blockchain, sync_options)
+			.map_err(|e| Error::Bdk(e))?;
+
+		Ok(())
+	}
+
+	pub(crate) fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync)>) -> Result<(), Error> {
+		let client = &*self.blockchain;
+
+		let cur_height = client.get_height()?;
+
+		let mut locked_last_sync_height = self.last_sync_height.lock().unwrap();
+		if cur_height >= locked_last_sync_height.unwrap_or(0) {
+			{
+				// First, inform the interface of the new block.
+				let cur_block_header = client.get_header(cur_height)?;
+				for c in &confirmables {
+					c.best_block_updated(&cur_block_header, cur_height);
+				}
+
+				*locked_last_sync_height = Some(cur_height);
+			}
+
+			{
+				// First, check the confirmation status of registered transactions as well as the
+				// status of dependent transactions of registered outputs.
+				let mut locked_queued_transactions = self.queued_transactions.lock().unwrap();
+				let mut locked_queued_outputs = self.queued_outputs.lock().unwrap();
+				let mut locked_watched_transactions = self.watched_transactions.lock().unwrap();
+				let mut locked_watched_outputs = self.watched_outputs.lock().unwrap();
+
+				let mut confirmed_txs = Vec::new();
+
+				// Check in the current queue, as well as in registered transactions leftover from
+				// previous iterations.
+				let mut registered_txs: Vec<Txid> = locked_watched_transactions
+					.iter()
+					.chain(locked_queued_transactions.iter())
+					.cloned()
+					.collect();
+
+				registered_txs.sort_unstable_by(|txid1, txid2| txid1.cmp(&txid2));
+				registered_txs.dedup_by(|txid1, txid2| txid1.eq(&txid2));
+
+				// Remember all registered but unconfirmed transactions for future processing.
+				let mut unconfirmed_registered_txs = Vec::new();
+
+				for txid in registered_txs {
+					if let Some(tx_status) = client.get_tx_status(&txid)? {
```

Comment:
Yep, they are all just requests directly hitting the Esplora server for now. While these calls are kind of specific to get the current status of a transaction, we may want to look into using the BDK wallet cache to retrieve some data when we tackle #3 .

Reply:
Well, Esplora doesn't support any kind of batching really, but we could at least make some of the functionality run concurrently? As it stands I grouped the method into three scopes anyways (essentially corresponding to `best_block_updated`, `transactions_confirmed`, `transaction_unconfirmed`). Could these three run in separate threads without an issue? The `Confirm::transactions_confirmed` docs say: "However, in the event of a chain reorganization, it must not be called with a header that is no longer in the chain as of the last call to [best_block_updated](https://docs.rs/lightning/latest/lightning/chain/trait.Confirm.html#tymethod.best_block_updated)."
But how bad would it be to let all three concurrently rather than in order?

Also, I could take the road of BDK's wallet sync and spawn a number of concurrent threads for each client request.

Reply:
> I understood that we were doing things async, does BDK not support rust-async esplora requests? Spinning up more threads probably isn't worth it, but if we can do parallel requests by just polling multiple futures together that likely is.

No, currently all is blocking. `rust-esplora-client` supports async, but using this would complicate things quite a bit, IIUC. For example,  we'd need to run `LdkLiteChainAccess` in its dedicated thread and then start communicating with it, e.g., via message passing. Happy to explore this further in a follow-up, but to keep things a bit more simple, I'd like to keep it blocking for now.
 
> More generally, I wonder seriously about what happens if we get a reorg in the middle of this function, or between calls for a single transaction. I didn't carefully review it, but at first glance I'm a bit dubious that we meet the LDK requirements of topological sorting of transactions (and block-disconnections) in such a case.

Mh, are there any (undocumented) requirements that I'm not aware of? I tried to consider the following:

1) Notify of new blocks, if there are intermediaries, feeding just the newest one is fine.
2) Notify of newly confirmed transactions, ensure inter- and intra-block ordering.
3) Notify of newly unconfirmed transactions.

Reply:
Btw. one architectural drawback of going async is that we will need to pass the tokio runtime around, e.g., to use the async client in #21. That on the other hands means that we have to setup the runtime before initializing `LdkLiteChainAccess`, i.e., in the builder, not only upon `start()`. This is a pity since I had imagined that we could just discard the `LdkLiteRuntime` on `stop()` and would get automatically get rid of any background tasks etc.